test/jdk/java/lang/module/ModuleDescriptorTest.java

```diff
@@ -1522,4 +1524,62 @@ public void testToString() {
         assertTrue(s.contains("p1"));
     }
 
+    @Test(expectedExceptions = InvalidModuleDescriptorException.class)
+    public void testRequiresTransitiveJavaBaseNotPermitted1() throws Exception {
+        ModuleDescriptor descriptor = ModuleDescriptor.newModule("foo")
+                .requires(Set.of(Modifier.TRANSITIVE), "java.base")
+                .build();
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ModuleInfoWriter.write(descriptor, baos);
+        ByteBuffer bb = ByteBuffer.wrap(baos.toByteArray());
+
+        ModuleDescriptor.read(bb, () -> Set.of("p", "q"));
+    }
+
+    @Test(expectedExceptions = InvalidModuleDescriptorException.class)
+    public void testRequiresTransitiveJavaBaseNotPermitted2() throws Exception {
+        ModuleDescriptor descriptor = ModuleDescriptor.newModule("foo")
+                .requires(Set.of(Modifier.TRANSITIVE), "java.base")
+                .build();
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ModuleInfoWriter.write(descriptor, baos);
+        byte[] bytecode = baos.toByteArray();
+        bytecode = setClassFileVersion(ClassFile.JAVA_21_VERSION, -1, bytecode);
+        ByteBuffer bb = ByteBuffer.wrap(bytecode);
+
+        ModuleDescriptor.read(bb, () -> Set.of("p", "q"));
+    }
+
+    public void testRequiresTransitiveJavaBasePermitted() throws Exception {
+        ModuleDescriptor descriptor = ModuleDescriptor.newModule("foo")
+                .requires(Set.of(Modifier.TRANSITIVE), "java.base")
+                .build();
+
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ModuleInfoWriter.write(descriptor, baos);
+        byte[] bytecode = baos.toByteArray();
+        bytecode = setClassFileVersion(-1, ClassFile.PREVIEW_MINOR_VERSION, bytecode);
+        ByteBuffer bb = ByteBuffer.wrap(bytecode);
+
+        descriptor = ModuleDescriptor.read(bb, () -> Set.of("p", "q"));
+
+        assertEquals(descriptor.requires().size(), 1);
+        Requires javaBase = descriptor.requires().iterator().next();
+        assertEquals(javaBase.name(), "java.base");
+        assertEquals(javaBase.modifiers(), Set.of(Modifier.TRANSITIVE));
+    }
+
+    private byte[] setClassFileVersion(int major, int minor, byte[] bytecode) {
```

Comment:
Would it be possible to add a method description and re-format the overly long lines as it's currently impossible to look at the change side-by-side.